omp/distributed/partition_kernels.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/distributed/partition_kernels.hpp"
+
+
+#include <omp.h>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+namespace partition {
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+void build_starting_indices(std::shared_ptr<const DefaultExecutor> exec,
+                            const GlobalIndexType* range_offsets,
+                            const int* range_parts, size_type num_ranges,
+                            int num_parts, int& num_empty_parts,
+                            LocalIndexType* ranks, LocalIndexType* sizes)
+{
+    std::fill_n(sizes, num_parts, 0);
+    auto num_threads = static_cast<size_type>(omp_get_max_threads());
+    auto size_per_thread = (num_ranges + num_threads - 1) / num_threads;
+    vector<LocalIndexType> local_sizes(num_parts * num_threads, 0, {exec});
+#pragma omp parallel
+    {
```

Comment:
This kernel could be replaced by the same code as for cuda/hip. Thrust can, in fact, be used on the host with an openmp backend as well. You'll just have to look carefully at what Thrust macro to define where so that it uses OpenMP in this file.

I guess one issue is we don't look for Thrust when we're not building the cuda or hip backends. If we don't want to add a thrust dependency for that, perhaps it makes sense to keep this.

Reply:
Yes, I don't think the benefit from unifying it is bigger than the hassle from using/not using thrust, so I will keep it as it is.